internal/common/ingest/ingestion_pipeline.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/G-Research/armada/internal/common/eventutil"
+	"github.com/G-Research/armada/internal/pulsarutils"
+	"github.com/G-Research/armada/pkg/armadaevents"
+)
+
+// HasPulsarMessageIds should be implemented by structs that can store a batch of pulsar message ids
+// This is needed so we can pass message Ids down the pipeline and ack them at the end
+type HasPulsarMessageIds interface {
+	GetMessageIDs() []pulsar.MessageID
+}
+
+// InstructionConverter should be implemented by structs that can convert a batch of event sequences into an object
+// suitable for passing to the sink
+type InstructionConverter[T HasPulsarMessageIds] interface {
+	Convert(*EventSequencesWithIds) T
+}
+
+// Sink should be implemented by the struct responsible for putting the data in its final resting place, e.g. a
+// database.
+type Sink[T HasPulsarMessageIds] interface {
+	// Store should persist the sink.  The store is responsible for retrying failed attempts and should only return an error
+	// When it is satisfied that operation cannot be retries.
+	Store(T) error
+}
+
+// EventSequencesWithIds consists of a batch of Event Sequences along with the corresponding Pulsar Message Ids
+type EventSequencesWithIds struct {
+	EventSequences []*armadaevents.EventSequence
+	MessageIds     []pulsar.MessageID
+}
+
+// IngestionPipeline is an application tha reads message from pulsar and inserts them into a sink. The pipeline will
+// handle the following automatically:
+//   - Receiving messages from pulsar
+//   - Splitting messages into batches for efficient processing
+//   - Unmarshalling into event sequences
+//   - Acking processed messages
+//   - Gracefully terminating on a SIGTERM
+//
+// Callers must supply two structs, an InstructionConverter for converting event sequences into something that can be
+// exhausted and a Sink capable of exhausting these objects
+type IngestionPipeline[T HasPulsarMessageIds] struct {
+	pulsarConfig           configuration.PulsarConfig
+	metricsConfig          configuration.MetricsConfig
+	pulsarSubscriptionName string
+	pulsarbatchSize        int
+	pulsarbatchDuration    time.Duration
+	converter              InstructionConverter[T]
+	sink                   Sink[T]
+}
+
+func NewIngestionPipeline[T HasPulsarMessageIds](pulsarConfig configuration.PulsarConfig,
+	pulsarSubscriptionName string,
+	pulsarbatchSize int,
+	pulsarbatchDuration time.Duration,
+	converter InstructionConverter[T],
+	sink Sink[T],
+	metricsConfig configuration.MetricsConfig,
+) *IngestionPipeline[T] {
+	return &IngestionPipeline[T]{
+		pulsarConfig:           pulsarConfig,
+		metricsConfig:          metricsConfig,
+		pulsarSubscriptionName: pulsarSubscriptionName,
+		pulsarbatchSize:        pulsarbatchSize,
+		pulsarbatchDuration:    pulsarbatchDuration,
+		converter:              converter,
+		sink:                   sink,
+	}
+}
+
+// Run will run the ingestion pipeline until a SIGTERM is received
+func (ingester *IngestionPipeline[T]) Run() {
```

Comment:
It's better to pass in a context. That way the caller is responsible for closing this function. The contexts used throughout the app make up a tree. The root context is created in the main function and is already set up to cancel on sigterm. When the root context is cancelled, the cancellation propagates to all child contexts.

Reply:
done